feature/contributor/src/main/java/io/github/droidkaigi/confsched2020/contributor/ui/viewmodel/ContributorsViewModel.kt

```diff
@@ -1,11 +1,87 @@
 package io.github.droidkaigi.confsched2020.contributor.ui.viewmodel
 
+import androidx.lifecycle.LiveData
 import androidx.lifecycle.ViewModel
+import androidx.lifecycle.asLiveData
+import androidx.lifecycle.liveData
+import io.github.droidkaigi.confsched2020.ext.combine
+import io.github.droidkaigi.confsched2020.ext.toAppError
+import io.github.droidkaigi.confsched2020.ext.toLoadingState
+import io.github.droidkaigi.confsched2020.model.AppError
+import io.github.droidkaigi.confsched2020.model.Contributor
+import io.github.droidkaigi.confsched2020.model.LoadState
 import io.github.droidkaigi.confsched2020.model.repository.ContributorRepository
+import java.lang.Exception
 import javax.inject.Inject
 
 class ContributorsViewModel @Inject constructor(
     private val contributorRepository: ContributorRepository
 ) : ViewModel() {
-    // TODO Implement
+
+    /**
+     * [ViewState] has to be wrapped to workaround a compile error about type mismatch.
+     * [combine] doesn't seem to understand child classes.
+     */
+    data class UiModel(val viewState: ViewState) {
+        companion object {
+            val DEFAULT = UiModel(ViewState.Loading)
+        }
+    }
+
+    sealed class ViewState {
+        object Loading : ViewState()
+
+        data class Loaded(val contributors: List<Contributor>) : ViewState()
+
+        data class Error(val error: AppError?) : ViewState()
+    }
+
+    private var contributors: List<Contributor> = emptyList()
+
+    private val contributorsLoadStateLiveData: LiveData<LoadState<List<Contributor>>> = liveData {
+        emitSource(
+            contributorRepository.contributorContents()
+                .toLoadingState()
+                .asLiveData()
+        )
+
+        try {
+            contributorRepository.refresh()
+        } catch (exception: Exception) {
+            // Emit error so that uiModel knows when it's done loading
+            emit(toErrorLoadState(exception))
+        }
+    }
+
+    val uiModel: LiveData<UiModel> = combine(
+        initialValue = UiModel.DEFAULT,
+        liveData1 = contributorsLoadStateLiveData
+    ) { _, loadState ->
+        when (loadState) {
+            is LoadState.Loading -> UiModel(ViewState.Loading)
+            is LoadState.Error -> {
+                val appError = loadState.getErrorIfExists()?.toAppError()
+                if (contributors.isNotEmpty()) {
+                    UiModel(ViewState.Loaded(contributors))
```

Comment:
I wanted to introduce a `FullError` state where the screen would display a "Retry" button. This would be for the users who just installed the app and visited the Contributors page for the first time(no cache yet) when they happen to have bad network connection.

If you are interested, here is what I tried:

`
fun onRetry() {
    reloadContributorsLiveData.value = LoadingState.Loading
    viewModelScope.launch() {
        try {
            contributorRepository.refresh()
            reloadContributorsLiveData.value = LoadingState.Loaded
        } catch (exception: Exception) {
            reloadContributorsLiveData.value = LoadingState.Error(exception)
        }
    }
}
`

`reloadContributorsLiveData.value` is added as `liveData2` to the `uiModel`'s `combine(...)`. When `refresh()` executes, it doesn't trigger another event in `contributorsLoadStateLiveData `'s `emitSource`. I guess it's not possible to trigger an event to `emitSource` from outside the `liveData` builder function?

Reply:
👀

Reply:
It would be better to be able to retry!
I think we should add `reloadContributorsLiveData` as you commented. 👀 
I have an implementation image like this.

In fragment

```
viewModel.uiModel.observe { uiModel ->
  if(uiModel.hasError) {
    showSnackbar(uiModel.error.message) { viewMode.retry() }
  }
}
```

Reply:
Sounds good. I will look into it in a separate PR.